### PR TITLE
Adding Pin

### PR DESCRIPTION
**Pull request by cgentry7**
This adds the Pin class.

### CODE REVIEW

File: mpactpy/pin.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        The pin mesh associated with this pin
+    materials : List[Material]
+        The materials used in each pin XSR. The number of
+        entries must be equal to the number of uniform
+        material regions defined in the pin mesh
+    pitch : Pitch
+        The pitch of the pin in each axis direction (keys: 'X', 'Y', 'Z') (cm)
+    """
+
+    class Pitch(TypedDict):
```
Comment (npl436) [inline mpactpy/pin.py:35]: Why define this here?  It seems like if you actually need to define it (do you?), the right place to do it is in the pinmesh class.

File: mpactpy/utils.py [hunk 2]
```diff
@@ -33,3 +38,76 @@ def relative_round(value: float, rel_tol: float =1e-9) -> float:
...
+        if isinstance(num, float):
+            if math.isclose(num, round(num)):
+                return f"{num:.1f}" if print_length is None else f"{num:{print_length}.1f}"
+            return f"{num:.15g}" if print_length is None else f"{num:{print_length}.15g}"
+        return f"{str(num)}" if print_length is None else f"{str(num):{print_length}}"
+
+    return ' '.join(print_num(x, print_length) for x in input_list)
+
+T = TypeVar('T', bound=Hashable)
+def unique(elements: List[T]) -> List[T]:
```
Comment (npl436) [inline mpactpy/utils.py:100]: There is nothing wrong with this function, but is it really necessary?  To do the same thing, I would have done something like:
```
a = list(set(a))
```

File: test/unit/test_pin.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+def test_pin_write_to_string(pin, pinmesh, material):
+    output = pin.write_to_string(prefix="test ",
+                                 material_mpact_ids={material: 3},
+                                 pinmesh_mpact_ids={pinmesh: 2},
+                                 pin_mpact_ids={pin: 1})
+    expected_output = "test pin 1 2 / 3 3 3 3 3 3 3 3 3\n"
+    assert output == expected_output
+
+def test_pin_get_axial_slice(pin):
+    pin_slice = pin.get_axial_slice(0.5, 1.5)
```
Comment (npl436) [inline test/unit/test_pin.py:49]: To be complete, you could also write a separate test for invalid `start_pos` and `stop_pos` and assert the method returns None.

File: test/unit/test_pin.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    assert pin_slice.pinmesh.number_of_material_regions == 8
+    assert pin_slice.pinmesh.regions_inside_bounds == [0, 1, 2, 4, 5, 6]
+    assert_allclose([pin_slice.pinmesh.pitch[i] for i in ['X','Y','Z']], [2., 2., 1.])
+    assert_allclose(pin_slice.pinmesh.zvals, [0.5, 1.0])
+
+def test_pin_axial_merge(pin):
+    merged_pin = pin.axial_merge(pin)
```
Comment (npl436) [inline test/unit/test_pin.py:57]: Same as above.  You could have a test for invalid pinmeshes.

File: mpactpy/pin.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        Returns
+        -------
+        Pin
+            The new combined pin
+        """
+
+        pinmesh_template = deepcopy(self.pinmesh)
+        pinmesh_template.set_axial_mesh(pin.pinmesh.zvals, pin.pinmesh.ndivz)
+
+        assert pin.pinmesh == pinmesh_template, "Pin meshes of the two pins must have the same radial specifications"
```
Comment (npl436) [inline mpactpy/pin.py:170]: For `get_axial_slice` you return `None` if the `start_pos` and `stop_pos` are invalid.  Here you have a different behavior if the pinmeshes aren't valid.  Why not have both methods behave similarly?  Either `assert` or return `None` in both?

**Review by npl436 — APPROVED** [review]
Thank you for the well-documented code.  I will merge.  It is up to you to address my comments or not.

File: mpactpy/pin.py (hunk 1 above)
Comment (cgentry7) [reply mpactpy/pin.py:35]: Yeah, so this is somewhat a stylistic thing, but I'm doing this for better typehinting.  I could import this across multiple classes, but I think it's better to define it uniquely within each class to avoid accidental complexity (i.e. a needless coupling across classes) and limit it to a specific context.

File: mpactpy/utils.py (hunk 2 above)
Comment (cgentry7) [reply mpactpy/utils.py:100]: Just more readable and explicit as to the intent of the operation.  If someone put `list(set(a))` in front of me, I wouldn't immediately think to myself, "yes, they trying to remove list redundancies"

File: mpactpy/pin.py (hunk 5 above)
Comment (cgentry7) [reply mpactpy/pin.py:170]: I guess I think of "getting a slice of nothing" as a valid operation.  So it's not that `start_pos` and `stop_pos` are invalid, it's just that you got a slice of nothing, which is itself nothing.  I can imagine client code that would prefer this behavior, like trying to see if a pin exists at a given axial region (try to slice the pin at the axial plane and if `None`, then pin doesn't exist there).

With the axial merge, the two pins must have the same radial pinmesh, always.  We maybe in the future could implement something that unionizes the two meshes into a consistent radial mesh, but haven't done so here.  As such, if these two radial pinmeshes differ, the operation cannot be done.  I can't imagine any client code that would be cool with trying to call a merge operation and failing, hence the assertion.

File: test/unit/test_pin.py (hunk 3 above)
Comment (cgentry7) [reply test/unit/test_pin.py:49]: Added in #22

File: test/unit/test_pin.py (hunk 4 above)
Comment (cgentry7) [reply test/unit/test_pin.py:57]: Added in #22
